plugins/grounded/grounded.cpp

```diff
@@ -0,0 +1,177 @@
+#include "ProcessWindows.h"
+#define MUMBLE_PLUGIN_NO_DEFAULT_FUNCTION_DEFINITIONS
+#include "MumblePlugin.h"
+#undef MUMBLE_PLUGIN_NO_DEFAULT_FUNCTION_DEFINITIONS
+
+#include "mumble_positional_audio_utils.h"
+
+
+using GroundedHandle = std::tuple< ProcessWindows, procptr_t >;
+
+static std::unique_ptr< GroundedHandle > handle;
+
+/* the indexes of these float[3] from the game are
+ *   0: south low, north high
+ *   1: west low, east high
+ *   2: altitude, high toward sky, goes up when you jump */
+struct GroundedCam {
+	/* 940 */ float top[3];
+	uint8_t _unused1[4];
+	/* 950 */ float front[3];
+	uint8_t _unused2[4 * 40];
+	/* 9fc */ float pos[3];
+};
+
+#define unreal_to_mumble_units(unreal) (unreal / 100.f)
+
+static_assert(sizeof(struct GroundedCam) == 200, "");
+
+
+mumble_error_t mumble_init(uint32_t) {
+	return MUMBLE_STATUS_OK;
+}
+
+void mumble_shutdown() {
+}
+
+MumbleStringWrapper mumble_getName() {
+	static const char name[] = "Grounded";
+
+	MumbleStringWrapper wrapper;
+	wrapper.data           = name;
+	wrapper.size           = strlen(name);
+	wrapper.needsReleasing = false;
+
+	return wrapper;
+}
+
+MumbleStringWrapper mumble_getDescription() {
+	static const char description[] = "Positional audio support for Grounded. Steam release version 1.4.3.4578.";
+
+	MumbleStringWrapper wrapper;
+	wrapper.data           = description;
+	wrapper.size           = strlen(description);
+	wrapper.needsReleasing = false;
+
+	return wrapper;
+}
+
+MumbleStringWrapper mumble_getAuthor() {
+	static const char author[] = "MumbleDevelopers";
+
+	MumbleStringWrapper wrapper;
+	wrapper.data           = author;
+	wrapper.size           = strlen(author);
+	wrapper.needsReleasing = false;
+
+	return wrapper;
+}
+
+mumble_version_t mumble_getAPIVersion() {
+	return MUMBLE_PLUGIN_API_VERSION;
+}
+
+void mumble_registerAPIFunctions(void *) {
+}
+
+void mumble_releaseResource(const void *) {
+}
+
+mumble_version_t mumble_getVersion() {
+	return { 1, 0, 0 };
+}
+
+uint32_t mumble_getFeatures() {
+	return MUMBLE_FEATURE_POSITIONAL;
+}
+
+uint8_t mumble_initPositionalData(const char *const *programNames, const uint64_t *programPIDs, size_t programCount) {
+	for (size_t i = 0; i < programCount; ++i) {
+		if (strcmp(programNames[i], "Maine-Win64-Shipping.exe") != 0) {
+			continue;
+		}
+
+		ProcessWindows proc(programPIDs[i], programNames[i]);
+
+		if (!proc.isOk()) {
+			continue;
+		}
+
+		const Modules &modules = proc.modules();
+		const auto iter        = modules.find("Maine-Win64-Shipping.exe");
+
+		if (iter == modules.cend()) {
+			continue;
+		}
+
+		procptr_t p = iter->second.baseAddress();
+
+		if (!(p = proc.peekPtr(p + 0x0612cb38))) {
+			continue;
+		}
+
+		if (!(p = proc.peekPtr(p + 0x0))) {
+			continue;
+		}
+
+		if (!(p = proc.peekPtr(p + 0x8))) {
+			continue;
+		}
+
+		p += 0x8c0;

```

Comment:
These should have some comments as to what exactly is being skipped over :thinking:

Reply:
I don't really know what those addresses are pointing to. I don't have access to the source code and I'm not adept at reverse engineering 🙃. My guess is that it's some structures about the current game world and the current player or their camera. So I'm not sure what comment I could add that would be helpful. (I guess one thing I could mention is that the lookups are expected to fail if the player has yet to load in to a game.)

Reply:
Why are you doing multiple skips instead of a single, large one then? Doing it step by step seems to suggest some knowledge of what is being skipped :thinking:

Reply:
It's a readability/formatting thing. It was my intention to write it the following way; but clang-format appears to not permit this:

``` c++
		if (   !(p = proc.peekPtr(p + 0x0612cb38))
		    || !(p = proc.peekPtr(p + 0x0))
		    || !(p = proc.peekPtr(p + 0x8))
		    || !(p = proc.peek(p + 0x8c0, _cam))) {
			continue;
		}
```

clang-format instead wraps it as something like

``` c++
if (!(p = proc.peekPtr(p + 0x0612cb38)) || !(p = proc.peekPtr(p + 0x0)) ||
    !(p = proc.peekPtr(p + 0x8)) || !(p = proc.peek(p + 0x8c0, _cam))) {
    continue;
}
```

It seemed to me that separating it out was a bit nicer to ingest but it's totally subjective and it was not my intention to suggest knowledge of the structures in these lookups. So I'm totally willing to change this so the peeks are done in one if conditional if you prefer.

Reply:
No, what I am asking is. why aren't you e.g. doing a single skip of e.g. 100 bytes instead of 10 times 10 bytes :thinking:

Reply:
The code resolves a pointer chain. There are 3 pointers before reaching the relevant struct.

`0x8c0` is the offset inside said struct.

Reply:
Ah, I see.